src/Enemy/EnemyStateHackStart.cpp

```diff
@@ -0,0 +1,176 @@
+    : al::ActorStateBase("憑依開始", rootActor) {
+    if (param) {
+        mParam = param;
+    } else {
+        mParam = &fallbackParam;
+    }
+    initNerve(&DiveIn, 0);
+    mPlayerHackStartShaderCtrl = new PlayerHackStartShaderCtrl(rootActor, shaderParam);
+}
+
+IUsePlayerHack* EnemyStateHackStart::tryStart(const al::SensorMsg* sensor, al::HitSensor* source, al::HitSensor* target) {
+    if (!rs::isMsgStartHack(sensor))
+        return 0;
+    al::setVelocityZero(mActor);
+    mHackActor = rs::startHack(target, source, 0);
+    rs::startHackStartDemo(mHackActor, mActor);
+    al::setNerve(this, &DiveIn);
+    return mHackActor;
+}
+
+void EnemyStateHackStart::kill() {
+    mIsDead = true;
+    if (!mHackActor)
+        return;
+    rs::endHackStartDemo(mHackActor, mActor);
+    mHackActor = nullptr;
+}
+
+bool EnemyStateHackStart::isHackStart() const {
+    return al::isNerve(this, &HackStart);
+}
+
+float EnemyStateHackStart::calcHackStartNerveRate() const {
+    if (isHackStart()) {
+        s32 frameMax = al::getActionFrameMax(mActor, mParam->mActionName);
+        return al::calcNerveRate(this, frameMax);
+    }
+    return 0;
+}
+
+void EnemyStateHackStart::exeDiveIn() {
+    if (!rs::isHackStartDemoEnterMario(mHackActor))
+        return;
+    if (mParam->mActionName)
+        al::setNerve(this, &HackStart);
+    else
+        kill();
+}
+
+void EnemyStateHackStart::exeHackStart() {
+    if (al::isFirstStep(this)) {
+        mPlayerHackStartShaderCtrl->start();
+        al::startAction(mActor, mParam->mActionName);
+
+        if (mParam->mHasSubActors) {
+            s32 subActorNum = al::getSubActorNum(mActor);
+            for (s32 i = 0; i < subActorNum; i++) {
+                if (al::isExistAction(al::getSubActor(mActor, i), mParam->mActionName)) {
+                    al::startAction(al::getSubActor(mActor, i), mParam->mActionName);
+                }
+            }
+        }
+        if (mParam->mVisAnimName) {
+            al::startVisAnim(mActor, mParam->mVisAnimName);
+        }
+        if (mParam->mMtpAnimName) {
+            al::startMtpAnim(mActor, mParam->mMtpAnimName);
+        }
+        al::LiveActor* actor = mActor;
+        if (al::isExistDepthShadowMapCtrl(actor)) {
+            al::invalidateShadow(actor);
+            al::offDepthShadowModel(actor);
+            al::validateDepthShadowMap(actor);
+        }
+        if (mParam->mUpdateSubActorShadowMap) {
+            s32 subActorNum = al::getSubActorNum(mActor);
+            for (s32 i = 0; i < subActorNum; i++) {
+                al::LiveActor* subActor = al::getSubActor(mActor, i);
+                if (al::isExistDepthShadowMapCtrl(subActor)) {
+                    al::invalidateShadow(subActor);
+                    al::offDepthShadowModel(subActor);
+                    al::validateDepthShadowMap(subActor);
+                }
+            }
+        }
```

Comment:
It refuses to inline if I try to use it in this function. Not sure if there's anything I can do here.

Reply:
Just for testing, try annotating it with `inline` or the gnu-attribute for it to ignore/help the heuristics, just to see if it would still be equivalent.

Reply:
Closing as this one seems pretty grim. Free to re-open if you have any other suggestions on it. Possible that there's a third inlined function that's being used within both of them, but I doubt it a little.